uqcsbot/dominos_coupons.py

```diff
@@ -139,27 +157,88 @@ def _get_coupons(n: int, ignore_expiry: bool, keywords: List[str]) -> List[Coupo
             for coupon in coupons
             if any(coupon.keyword_matches(keyword) for keyword in keywords)
         ]
-    return coupons[:n]
 
+    if not source:
+        random.shuffle(coupons)
+        return coupons[:n]
 
-def _get_coupons_from_page() -> List[Coupon]:
-    """
-    Strips results from html page and returns a list of Coupon(s)
-    """
-    http_response = requests.get(COUPONESE_DOMINOS_URL)
-    if http_response.status_code != requests.codes.ok:
-        raise HTTPResponseException(http_response.status_code)
-    soup = BeautifulSoup(http_response.content, "html.parser")
-    soup_coupons = soup.find_all(class_="ov-coupon")
+    return coupons[:n]
 
+
+def _get_coupons_from_page(source: str) -> List[Coupon]:
+    urls: List[str] = []
     coupons: List[Coupon] = []
 
-    for soup_coupon in soup_coupons:
-        expiry_date_str = soup_coupon.find(class_="ov-expiry").get_text(strip=True)
-        description = soup_coupon.find(class_="ov-desc").get_text(strip=True)
-        code = soup_coupon.find(class_="ov-code").get_text(strip=True)
-        coupon = Coupon(code, expiry_date_str, description)
-        coupons.append(coupon)
+    website_coupon_classes: Dict[str, Dict[str, str]] = {
+        COUPONESE_DOMINOS_URL: {
+            "expiry": "ov-expiry",
+            "description": "ov-desc",
+            "code": "ov-code",
+        },
+        FRUGAL_FEEDS_DOMINOS_URL: {
+            "expiry": "column-3",
+            "description": "column-2",
+            "code": "column-1",
+        },
+    }
+
+    if source == "couponese":
+        urls.append(COUPONESE_DOMINOS_URL)
+    elif source == "frugalfeeds":
+        urls.append(FRUGAL_FEEDS_DOMINOS_URL)
+    else:
+        urls = [FRUGAL_FEEDS_DOMINOS_URL, COUPONESE_DOMINOS_URL]
+
+    for url in urls:
+        http_response: requests.Response = requests.get(url)
+        if http_response.status_code != requests.codes.ok:
+            raise HTTPResponseException(http_response.status_code, url)
+
+        soup = BeautifulSoup(http_response.content, "html.parser")
+        soup_coupons: List[BeautifulSoup] = []
+
+        if url == COUPONESE_DOMINOS_URL:
+            soup_coupons = soup.find_all(class_="ov-coupon")
+        elif url == FRUGAL_FEEDS_DOMINOS_URL:
+            tables = soup.select('[class^="tablepress"]')
+            for table in tables:
+                # Headers have stuff we don't want
+                rows = table.find_all("tr")[1:]
+                soup_coupons.extend(rows)
+
+        siteclass: Dict[str, str] = website_coupon_classes.get(url, {})
+
+        for soup_coupon in soup_coupons:
+            expiry_date_container = soup_coupon.find(class_=siteclass.get("expiry"))
+            description_container = soup_coupon.find(
+                class_=siteclass.get("description")
+            )
+            code_container = soup_coupon.find(class_=siteclass.get("code"))
+
+            if (
+                not expiry_date_container
```

Comment:
Potentially we might be able to still show coupons that don't have a date, but this is up to you.

Reply:
I'll have a look into this, however the user is generally provided enough coupon codes where we can get away with excluding coupons that don't have a date. (and I'm not sure if I've come across a coupon code without a date yet)